### ISSUE TITLE
Self Organizing Map (SOM) Package for Python (SOMPY) issue with njob > 1

### ISSUE THREAD

**Issue by jiperez91** [joblib/joblib#374]
Hi everyone:

I'm a student of Computer Engineering and I'm trying to use the Self Organizing Map (SOM) Package for Python (SOMPY). I want to speed up the algorithm (reducing the total time elapsed) and I see that in the code it uses Joblib. The train function has a parameter (njob) that I think it can help me. I have been studying the function and supposedly I have to put njob = 6 in order to use all the cores of my processor. The problem is when I call the train function with njob = 6, the algorithm is slower than with njob = 1. I've been reading this post http://stackoverflow.com/questions/21027477/joblib-parallel-multiple-cpus-slower-than-single but I still don't understand why.

This is the project: https://github.com/sevamoo/SOMPY

I REALLY need this algorithm to run faster.

I will be really grateful if you can help me with this problem.

Thank you for your time.

**Comment by lesteve** [comment]
Have you tried asking to the SOMPY people? They may have encountered similar issues in the past and possibly have some piece of advice for you.

**Comment by jiperez91** [comment]
Yes, I tried but they don't know either.

Look this simple example (it calculates the square root of numbers between 0 and 1000):

`
from math import sqrt
import time
from sklearn.externals.joblib import Parallel, delayed

if __name__ == "__main__":
    t_temp  = time.time()
    Parallel(n_jobs=2)(delayed(sqrt)(i) for i in xrange(1000))
    t_temp2 = time.time()
    print(t_temp2-t_temp)
`

Again, with njob > 1 is slower than njob = 1. So i don't understand how Joblib works.

**Comment by lesteve** [comment]
Readability counts, a lot! Please use triple back-quotes aka [fenced code blocks](https://help.github.com/articles/creating-and-highlighting-code-blocks/) to format error messages and code snippets.

I have updated your comment for this time, but try to remember it for next time.

**Comment by lesteve** [comment]
About your snippet, this is very likely because there is some overhead due to inter-process communication. Because computing a sqrt is very fast your computation time is completely dominated by this overhead.

```py
from math import sqrt
import time
from sklearn.externals.joblib import Parallel, delayed

def slow_sqrt(i):
    time.sleep(0.005)
    return sqrt(i)

if __name__ == "__main__":
    t_temp = time.time()
    Parallel(n_jobs=1)(delayed(slow_sqrt)(i) for i in range(1000))
    t_temp2 = time.time()
    print('n_jobs=1:', t_temp2-t_temp)

    t_temp = time.time()
    Parallel(n_jobs=2)(delayed(slow_sqrt)(i) for i in range(1000))
    t_temp2 = time.time()
    print('n_jobs=2:', t_temp2-t_temp)
```

```
n_jobs=1: 5.227390289306641
n_jobs=2: 2.726900577545166
```

Note the story is a little bit more complicated than this because joblib automatically does some batching of tasks to try to alleviate this problem. In your case above though this mechanism doesn't have time to do it enough. I tried this:

```py
from math import sqrt
import time
from sklearn.externals.joblib import Parallel, delayed

if __name__ == "__main__":
    t_temp = time.time()
    Parallel(n_jobs=1)(delayed(sqrt)(i) for i in range(100000))
    t_temp2 = time.time()
    print('n_jobs=1:', t_temp2-t_temp)

    t_temp = time.time()
    Parallel(n_jobs=2)(delayed(sqrt)(i) for i in range(100000))
    t_temp2 = time.time()
    print('n_jobs=2:', t_temp2-t_temp)
```

And the auto-batching mechanism has more time to batch tasks in large chunks so that n_jobs=2 is faster:
```
n_jobs=1: 3.624116897583008
n_jobs=2: 1.0857858657836914
```

**Comment by jiperez91** [comment]
In the last example I get:

```
('n_jobs=1:', 4.928999900817871)
('n_jobs=2:', 45.94200015068054)
```

I don't know why you get 1.085 seconds with n_jobs = 2

**Comment by lesteve** [comment]
Hmmm, what kind of timings do you get in the first example?

Could you give me the output of this snippet?

```py
import platform; print(platform.platform())
import sys; print("Python", sys.version)
from sklearn.externals import joblib; print(joblib.__version__)
import multiprocessing; print(multiprocessing.cpu_count())
```

**Comment by jiperez91** [comment]
In the first example, the timings are very similar (faster with n_job=2).

The output of the last is:

```
Windows-7-6.1.7600
('Python', '2.7.10 |Anaconda 2.3.0 (64-bit)| (default, May 28 2015, 16:44:52) [MSC v.1500 64 bit (AMD64)]')
0.8.4
6
```

**Comment by lesteve** [comment]
You have an old joblib that doesn't do the automatic batching. It was included in 0.9.2.

Update scikit-learn and you'll get a more up-to-date joblib version.

Note that you don't need to install scikit-learn if you use only joblib. You can install joblib with pip and import it directly instead of importing from sklearn.externals.

**Comment by ogrisel** [comment]
I think @lesteve is the right answer. Closing this issue. Please feel free to reopen with a comment if this is not the case for you.
